vertical-pod-autoscaler/pkg/recommender/routines/safety_margin_modifier_post_processor.go

```diff
@@ -0,0 +1,221 @@
+		case exponential:
+			if len(resourceModifier.Parameters) != 2 {
+				klog.Errorf("Skipping %s safety margin modifier in vpa %s/%s: exponential modifier requires 2 parameters, %d given", resourceName, name, vpa.ID.Namespace, vpa.ID.VpaName, len(resourceModifier.Parameters))
+				continue
+			}
+			newRec = exponentialModifier(recCopy, r.DefaultSafetyMarginFactor, resourceModifier.Parameters[0], resourceModifier.Parameters[1])
+		case undefined:
+			continue
+		default:
+			klog.Errorf("Skipping %s safety margin modifier in vpa %s/%s: specified modifier is not valid", name, vpa.ID.Namespace, vpa.ID.VpaName)
+		}
+
+		if resourceName == v1.ResourceCPU {
+			containerRec[resourceName] = *resource.NewMilliQuantity(int64(newRec), resource.DecimalSI)
+		} else {
+			containerRec[resourceName] = *resource.NewQuantity(int64(newRec), resource.DecimalSI)
+		}
+	}
+}
+
+// Undo default safety margin before applying a custom linear safety margin (a * baseRec)
+func linearModifier(rec resource.Quantity, defaultSafetyMargin, slope float64) float64 {
+	return float64(rec.Value()) / defaultSafetyMargin * slope
+}
+
+// Undo default safety margin before applying a custom affine safety margin (a * baseRec + b)
+func affineModifier(rec resource.Quantity, defaultSafetyMargin, constant, slope float64) float64 {
+	return float64(rec.Value())/defaultSafetyMargin*slope + constant
+}
+
+// Undo default safety margin before applying a custom logarithmic safety margin (baseRec + a * log10(baseRec))
+func logModifier(rec resource.Quantity, defaultSafetyMargin, factor float64) float64 {
+	rawRec := float64(rec.Value()) / defaultSafetyMargin
+	return rawRec + math.Log10(rawRec)*factor
+}
+
+// Undo default safety margin before applying a custom exponential safety margin (baseRec + a * baseRec^b)
+func exponentialModifier(rec resource.Quantity, defaultSafetyMargin, exponent, factor float64) float64 {
+	rawRec := float64(rec.Value()) / defaultSafetyMargin
+	return rawRec + math.Pow(rawRec, exponent)*factor
+}
+
+func readSafetyMarginModifierFromVPAAnnotations(vpa *model.Vpa) map[string]safetyMarginModifier {
+	modifiers := map[string]safetyMarginModifier{}
+	for key, value := range vpa.Annotations {
+		containerName := extractContainerName(key, vpaPostProcessorPrefix, vpaPostProcessorSafetyMarginModifierSuffix)
+		if containerName == "" {
+			continue
+		}
+
+		safetyMarginModifier := safetyMarginModifier{}
+		safetyMarginFunction := safetyMarginFunction{}
+		if err := json.Unmarshal([]byte(value), &safetyMarginModifier); err != nil {
+			if err := json.Unmarshal([]byte(value), &safetyMarginFunction); err != nil {
+				klog.Errorf("Skipping safety margin modifier definition '%s' for container '%s' in vpa %s/%s due to bad format, error:%#v", value, containerName, vpa.ID.Namespace, vpa.ID.VpaName, err)
+				continue
+			}
+			safetyMarginModifier[v1.ResourceCPU] = safetyMarginFunction
```

Comment:
I think you need to iterate on controlled resources instead no ?

Reply:
My undertstanding is that the iteration on resource is done by the first unmarshall attempt into `safetyMarginModifier` that is the struct indexed by resourceName.
Here we have the default that affect the function to all resources.

While it should work I find this code a bit hard to read. Maybe to have a simpler code and a uniq/consistent way of describing things in the API we should keep only the API where the user define the resource. In other words the one with `safetyMarginModifier`.

Reply:
When I started this, I was thinking it as just a way to override the default safety margin, so having the same modifier applying to CPU and Memory made more sense. But actually, the more I think about it, especially while writing the tests, the more I think that safety margin should indeed be different between CPU and memory by default. 

Though, I still like the possibility of being able to have wildcard modifier that applies to all resources, so I think I'll just check if the resource name is `*` and if so, apply the same modifier for CPU and memory.